pkg/util/allocationfilterutil/v2/lexer.go

```diff
@@ -0,0 +1,272 @@
+package allocationfilterutil
+
+import (
+	"fmt"
+
+	multierror "github.com/hashicorp/go-multierror"
+
+	"github.com/kubecost/opencost/pkg/kubecost"
+)
+
+// ============================================================================
+// This file contains:
+// Lexing (string -> []token) for V2 of allocation filters
+// ============================================================================
+//
+// See parser.go for a formal grammar and external links.
+
+type tokenKind int
+
+const (
+	colon tokenKind = iota // ':'
+	comma                  // ','
+	plus                   // '+'
+
+	bangColon // '!:'
+
+	str // '"foo"'
+
+	filterField1 // 'namespace', 'cluster'
+	filterField2 // 'label', 'annotation'
+	keyedAccess  // '[app]', '[foo]', etc.
+	identifier   // K8s valid name + sanitized Prom: 'app', 'abc_label'
+
+	eof
+)
+
+// These maps serve a dual purpose. (1) to help the lexer identify special
+// strings that should become filterField1/2 instead of identifiers and (2) to
+// help the parser convert tokens into AllocationFilterConditions.
+//
+// TODO: Should the lexer _also_ attach the FilterField to help the parser out?
```

Comment:
Essentially, the lexer could attach the appropriate `FilterField` to the lexed `token` (because it knows at lex time) instead of making the parser do an extra map lookup. I leaned away from this idea because it would complicate the `token` type, but I'm open to ideas!

Reply:
Trust you to make the call here, but my gut says that it's fine how it is. No need to make a tradeoff that increases complexity unless we're afraid this will be too slow, which seems unlikely. Maybe others will have opinions.